lib/xml/xml.go

```diff
@@ -163,10 +163,24 @@ func builtinTypeFor(typ xsd.Type) xsd.Builtin {
 func Unmarshal(r io.Reader, details map[string]Detail) (cdata string, elems map[string]any, err error) {
 	dec := xml.NewDecoder(r)
 	dec.CharsetReader = func(_ string, input io.Reader) (io.Reader, error) { return input, nil }
-	return walkXML(dec, nil, details)
+	var w walker
+	cdata, elems, err = w.walkXML(dec, nil, details)
+	if err == nil && !w.wasValidXML() {
+		err = io.ErrUnexpectedEOF
+	}
+	return cdata, elems, err
+}
+
+type walker struct {
+	hasDecl bool
+	hasElem bool
 }
 
-func walkXML(dec *xml.Decoder, attrs []xml.Attr, details map[string]Detail) (cdata string, elems map[string]any, err error) {
+func (w *walker) wasValidXML() bool {
+	return w.hasDecl && w.hasElem
```

Comment:
Not sure whether this should be as harsh as it is. If it were the disjunction, then otherwise valid docs without an XML declaration would not require a `try` wrapper.

Reply:
I'm thinking perhaps just returning an error if `len(elems) == 0`, that is pretty much what `hasElem` captures. This allows us to ingest docs that are missing an XML declaration.